### PR TITLE
GFM autolink extension

### LINKED ISSUE #201: Github Flavored Markdown extended autolinks unsupported
[The Github Flavored Markdown spec extends CommonMark's autolinks to support links without `<...>` surrounding them.](https://github.github.com/gfm/#autolinks-extension-)

It appears that the current GFM extension set does not include support for this.
image

Raw links were passed on before in #33, but that was due to them [not being a part of the CommonMark spec](http://spec.commonmark.org/0.24/#autolinks).

Would this project be open to a PR to add support for this?

### PR DESCRIPTION
**Pull request by dudeofawesome**
This is a very much a work in progress. 

Currently, I'm trying to figure out why a few of the tests are failing when there are multiple links to match. It looks like the RegExp only matches the first link. I'm also seeing some inconsistency with scheme-less link matching.

The CommonMark Autolink tests that are failing should be failing with GFM Autolink Extensions enabled.

This will fix #201

### CODE REVIEW

**Comment by dudeofawesome** [comment]
I signed it!

File: tool/gfm_tests.json [hunk 1]
```diff
@@ -5461,7 +5461,7 @@
     "example": 605,
     "start_line": 9025,
     "end_line": 9029,
-    "section": "Autolinks (extension)",
+    "section": "Autolinks Email (extension)",
```
Comment (kevmoo) [inline tool/gfm_tests.json]: FYI: this is generated from the GFM source. Shouldn't hand edit.

Comment (dudeofawesome) [reply tool/gfm_tests.json]: Alright, I can definitely revert that. It was just for testing to cut down on the number of tests being run.

Comment (kevmoo) [reply tool/gfm_tests.json]: Please! I just checked the latest GFM output – these haven't changed

Comment (dudeofawesome) [reply tool/gfm_tests.json]: I don't think that this is quite ready to merge yet, so I'd like to leave the modified tests as is for now to ease my dev cycle.

I'm still trying to figure out what is causing it to fail some of the Autolink (extension) tests. I'm not seeing why it's failing [example-598](https://github.com/dudeofawesome/dart-lang-markdown/blob/0a503e882d593f6402b215de9e0712e7b40183f8/tool/gfm_tests.json#L5382) and [example-599](https://github.com/dudeofawesome/dart-lang-markdown/blob/0a503e882d593f6402b215de9e0712e7b40183f8/tool/gfm_tests.json#L5393). I also am not sure how to fix [example-604](https://github.com/dudeofawesome/dart-lang-markdown/blob/0a503e882d593f6402b215de9e0712e7b40183f8/tool/gfm_tests.json#L5448).

**Comment by srawlins** [comment]
Woohoo I see only one failed test in Travis!

File: tool/common_mark_stats.txt [hunk 2]
```diff
@@ -24,4 +24,4 @@
   11 of   11 – 100.0%  Tabs
    3 of    3 – 100.0%  Textual content
   19 of   19 – 100.0%  Thematic breaks
- 561 of  624 –  89.9%  TOTAL
+ 558 of  624 –  89.4%  TOTAL
```
Comment (kevmoo) [inline tool/common_mark_stats.txt]: ooo...regression here!

File: tool/gfm_tests.json (hunk 1 above)
Comment (kevmoo) [reply tool/gfm_tests.json]: Please revert the changes to this file. It's autogenerated...

File: tool/common_mark_stats.json [hunk 3]
```diff
@@ -279,12 +279,12 @@
   "115": "fail"
  },
  "Hard line breaks": {
-  "605": "strict",
```
Comment (kevmoo) [inline tool/common_mark_stats.json]: I think you fixed these, too....looking at travis failures. May want to regenerate

File: lib/src/extension_set.dart [hunk 4]
```diff
@@ -35,11 +35,12 @@ class ExtensionSet {
     const FencedCodeBlockSyntax(),
     const HeaderWithIdSyntax(),
     const SetextHeaderWithIdSyntax(),
-    const TableSyntax()
+    const TableSyntax(),
```
Comment (kevmoo) [inline lib/src/extension_set.dart]: please avoid unrelated changes...just makes the diffs easier to read

File: lib/src/extension_set.dart [hunk 5]
```diff
@@ -48,10 +49,11 @@ class ExtensionSet {
   /// [GitHub flavored Markdown]: https://github.github.com/gfm/
   static final ExtensionSet gitHubFlavored = new ExtensionSet([
     const FencedCodeBlockSyntax(),
-    const TableSyntax()
+    const TableSyntax(),
```
Comment (kevmoo) [inline lib/src/extension_set.dart]: ditto

File: tool/common_mark_stats.txt (hunk 2 above)
Comment (kevmoo) [reply tool/common_mark_stats.txt]: Gotta get rid of these regressions...

Comment (dudeofawesome) [reply tool/common_mark_stats.txt]: Whoops, I was only generating the test files for GFM

File: tool/gfm_stats.json [hunk 6]
```diff
@@ -34,21 +34,21 @@
   "589": "strict",
   "590": "loose",
   "591": "loose",
-  "592": "loose",
```
Comment (kevmoo) [inline tool/gfm_stats.json]: Trading failures is less than ideal...but our total numbers are better.

Comment (dudeofawesome) [reply tool/gfm_stats.json]: If you look at the test cases for 592 and 595, you'll see that there's an incompatibility between the GFM autolink extensions and the original CM autolinks. 

Example 592 specifies that `< http://foo.bar >` should not get matched as a link, but the GFM autolink extension doesn't specify anything that would prevent that from matching.

Example 595 is literally just a link `http://example.com`, which obviously should be found by GFM autolinks.

Comment (dudeofawesome) [reply tool/gfm_stats.json]: On that note, I'm authoring an issue right now on the [github/cmark](https://github.com/github/cmark) repo to address this. 

@srawlins You had mentioned offline that there were a few other projects using the cmark-gfm examples as test cases. Can you link me to those?

Comment (srawlins) [reply tool/gfm_stats.json]: Er, I thought pulldown-cmark (Rust) was... but they appear to be [using regular CommonMark](https://github.com/google/pulldown-cmark/blob/8b211fde18df2b974163716189cd7b197bdbc6b4/src/passes.rs#L100) tests. I'm not sure of others. In any case, I think it would be hard for them to argue that the examples should remain.

Comment (srawlins) [reply tool/gfm_stats.json]: @kevmoo I'd like to comment somewhere that 2 tests are failing because of a bug in github/cmark's tests, but not sure where to comment on that... maybe we can all just remember in our brains.

**Comment by kevmoo** [comment]
Honestly, since we're careful about diffs in the generated output, I think
just putting the comment in the txt summary file is okay

WDYT?

File: tool/gfm_stats.json (hunk 6 above)
Comment (dudeofawesome) [reply tool/gfm_stats.json]: [Here's the issue I opened with github/cmark](https://github.com/github/cmark/issues/81).

File: lib/src/inline_parser.dart [hunk 7]
```diff
@@ -274,6 +274,118 @@ class AutolinkSyntax extends InlineSyntax {
   }
 }
 
+/// Matches autolinks like `http://foo.com`.
+class AutolinkExtensionSyntax extends InlineSyntax {
+  static const START = r'(?:^|[\s*_~(>])';
```
Comment (srawlins) [inline lib/src/inline_parser.dart]: Dart has moved to lowerSnakeCase for const stuff. Could you update these variable names?

File: lib/src/inline_parser.dart [hunk 8]
```diff
@@ -274,6 +274,118 @@ class AutolinkSyntax extends InlineSyntax {
...
+  static const SCHEME = r'(?:(?:https?|ftp):\/\/|www\.)';
+  static const DOMAIN_PART = r'\w\-';
+  static const DOMAIN = '[$DOMAIN_PART][$DOMAIN_PART.]+';
+  static const PATH = r'[^\s<]*';
+  // static const TRUNCATING_PUNCTUATION_NEG = r'[^\s<\?\!\.\,\:\*\_\~]';
```
Comment (srawlins) [inline lib/src/inline_parser.dart]: Delete.

File: lib/src/inline_parser.dart (hunk 7 above)
Comment (srawlins) [inline lib/src/inline_parser.dart]: Could you comment on what the static strings are, just as a group.

File: lib/src/inline_parser.dart [hunk 9]
```diff
@@ -274,6 +274,118 @@ class AutolinkSyntax extends InlineSyntax {
...
+  static const TRUNCATING_PUNCTUATION_NEG = r'';
```
Comment (srawlins) [inline lib/src/inline_parser.dart]: Could you put a specific comment for each of these strings. What is a "truncating punctuation"? You could even just mention to look at the comment for `trailingPunc` below.

File: lib/src/inline_parser.dart [hunk 10]
```diff
@@ -274,6 +274,118 @@ class AutolinkSyntax extends InlineSyntax {
...
+  static const TRUNCATING_PUNCTUATION_POS = r'[?!.,:*_~]';
+
+  AutolinkExtensionSyntax()
+      : super('$START(($SCHEME)($DOMAIN)($PATH))$TRUNCATING_PUNCTUATION_NEG');
+
+  @override
+  bool tryMatch(InlineParser parser) {
+    var startMatch = pattern.matchAsPrefix(
+        parser.source, (parser.pos > 0 ? parser.pos - 1 : 0));
+    if (startMatch != null) {
```
Comment (srawlins) [inline lib/src/inline_parser.dart]: Let's flip this logic, for less nesting:

```dart
if (startMatch == null) return false;

parser.writeText();
...
```

File: lib/src/inline_parser.dart [hunk 11]
```diff
@@ -274,6 +274,118 @@ class AutolinkSyntax extends InlineSyntax {
...
+    }
+
+    /**
+     * Trailing punctuation (specifically, ?, !, ., ,, :, *, _, and ~) will
+     * not be considered part of the autolink, though they may be included
+     * in the interior of the link:
+     * https://github.github.com/gfm/#example-599
+     */
+    final trailingPunc =
+        new RegExp('$TRUNCATING_PUNCTUATION_POS*' + r'$').firstMatch(url);
```
Comment (srawlins) [inline lib/src/inline_parser.dart]: Can you make this RegExp static as well, so that its not compiled every go.

File: lib/src/inline_parser.dart [hunk 12]
```diff
@@ -274,6 +274,118 @@ class AutolinkSyntax extends InlineSyntax {
...
+    /**
+     * When an autolink ends in ), we scan the entire autolink for the total
+     * number of parentheses. If there is a greater number of closing
+     * parentheses than opening ones, we don’t consider the last character
+     * part of the autolink, in order to facilitate including an autolink
+     * inside a parenthesis:
+     * https://github.github.com/gfm/#example-600
+     */
+    if (url.endsWith(')')) {
+      final opening = new RegExp(r'\(').allMatches(url).length;
```
Comment (srawlins) [inline lib/src/inline_parser.dart]: Let's not use RegExp just to count the number of parens. Can you write a private helper method that just walks the url and counts opening and closing? Should be much faster.

File: lib/src/inline_parser.dart [hunk 13]
```diff
@@ -274,6 +274,118 @@ class AutolinkSyntax extends InlineSyntax {
...
+      if (entityRef != null) {
+        // Strip out HTML entity reference
+        url = url.substring(0, url.length - entityRef[0].length);
+        href = href.substring(0, href.length - entityRef[0].length);
+        matchLength -= entityRef[0].length;
+      }
+    }
+
+    /** The scheme http will be inserted automatically */
+    if (!href.startsWith(new RegExp(r'(?:https?|ftp)\:\/\/'))) {
```
Comment (srawlins) [inline lib/src/inline_parser.dart]: Let's just use some strings, instead of a RegExp:

```dart
if (!href.startsWith('http://') &&
    !href.startsWith('https://') &&
    !href.startsWith('ftp://'))
```

File: lib/src/inline_parser.dart [hunk 14]
```diff
@@ -274,6 +274,118 @@ class AutolinkSyntax extends InlineSyntax {
...
+      href = 'http://$href';
+    }
+
+    final anchor = new Element.text('a', escapeHtml(url));
+    anchor.attributes['href'] = Uri.encodeFull(href);
+    parser.addNode(anchor);
+
+    parser.consume(matchLength);
+    return false;
```
Comment (srawlins) [inline lib/src/inline_parser.dart]: Why are we returning false here?

File: lib/src/inline_parser.dart (hunk 7 above)
Comment (dudeofawesome) [reply lib/src/inline_parser.dart]: What is the preferred comment style for Dart? I see a lot of `///` elsewhere in the document, but a few `//` as well.

Also, I'll update my other comments to match the preferred style.

File: lib/src/inline_parser.dart (hunk 10 above)
Comment (dudeofawesome) [reply lib/src/inline_parser.dart]: This is actually almost identical to the `super.tryMatch`, which has the logic in the current order. Should we change it in both places?

File: lib/src/inline_parser.dart (hunk 12 above)
Comment (dudeofawesome) [reply lib/src/inline_parser.dart]: Could I switch this to `url.allMatches('(').length;` for a similar speed improvement, or does this still use a RegExp underneath? The documentation doesn't seem to explicitly say that it does.

File: lib/src/inline_parser.dart (hunk 14 above)
Comment (dudeofawesome) [reply lib/src/inline_parser.dart]: `onMatch` returns whether the caller should advance the parser by `match[0].length`. Since I'm calling `parser.consume(matchLength);` above, it needs to not consume the match length again.

**Comment by dudeofawesome** [comment]
@kevmoo `dart tool/stats.dart --update-files` only updates the CM stats files. Is that intentional? If so, is there a way to run both CM and GFM in the same command? That's been throwing me off a few times.

**Comment by srawlins** [comment]
@dudeofawesome you can run `dart tool/stats.dart --update-files --flavor gfm`

File: lib/src/inline_parser.dart (hunk 14 above)
Comment (srawlins) [reply lib/src/inline_parser.dart]: OK sg.

File: lib/src/inline_parser.dart (hunk 10 above)
Comment (srawlins) [reply lib/src/inline_parser.dart]: Ah, I see, unfortunate that you have to override just for that little backtracking offset.

Not the end of the world though. You can leave the super.tryMatch logic alone. I might clean it up later.

File: tool/gfm_stats.json (hunk 6 above)
Comment (srawlins) [reply tool/gfm_stats.json]: Thanks that works.

**Comment by dudeofawesome** [comment]
Just did, no changes to commit.

File: lib/src/inline_parser.dart (hunk 10 above)
Comment (dudeofawesome) [reply lib/src/inline_parser.dart]: Perhaps I could add an optional offset parameter to `super.tryMatch`, and then just call that, so that I can remove all of the redundant code here?

Comment (srawlins) [reply lib/src/inline_parser.dart]: Yeah that's probably not too unruly. Let's see how it looks.

**Comment by kevmoo** [comment]
Of course I just messed it up by merging in the changes from @srawlins 

Sorry, @dudeofawesome – one more time?

**Comment by dudeofawesome** [comment]
Rebased and updated!

File: lib/src/inline_parser.dart [hunk 15]
```diff
@@ -274,6 +275,128 @@ class AutolinkSyntax extends InlineSyntax {
...
+  static final regExpTrailingPunc =
+      new RegExp('$truncatingPunctuationPositive*' + r'$');
+  static final regExpEndsWithColon = new RegExp(r'\&[a-zA-Z0-9]+;$');
+  static final regExpWhiteSpace = new RegExp(r'\s');
+
+  AutolinkExtensionSyntax() : super('$start(($scheme)($domain)($path))');
+
+  @override
+  bool tryMatch(InlineParser parser, [int startMatchPos]) {
+    return super.tryMatch(parser, parser.pos > 0 ? parser.pos - 1 : 0);
```
Comment (srawlins) [inline lib/src/inline_parser.dart]: Yeah this is very nice now.

**Comment by srawlins** [comment]
Code looks good; Travis is happy; GFM compliance up from 88% to **89%** Woohoo  🙌  🎉 🍾  ! Thanks much @dudeofawesome!
